crates/interpreter/src/instructions/system.rs

```diff
@@ -125,21 +125,46 @@ pub fn returndatasize<H: Host + ?Sized, SPEC: Spec>(interpreter: &mut Interprete
 pub fn returndatacopy<H: Host + ?Sized, SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut H) {
     check!(interpreter, BYZANTIUM);
     pop!(interpreter, memory_offset, offset, len);
+
     let len = as_usize_or_fail!(interpreter, len);
     gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
+
     let data_offset = as_usize_saturated!(offset);
-    let data_end = data_offset.saturating_add(len);
-    if data_end > interpreter.return_data_buffer.len() {
+    let return_data_buffer_len = interpreter.return_data_buffer.len();
+
+    if data_offset.saturating_add(len) > return_data_buffer_len && !interpreter.is_eof {
         interpreter.instruction_result = InstructionResult::OutOfOffset;
         return;
     }
-    if len != 0 {
-        let memory_offset = as_usize_or_fail!(interpreter, memory_offset);
-        resize_memory!(interpreter, memory_offset, len);
+
+    if len == 0 {
+        return;
+    }
+
+    let memory_offset = as_usize_or_fail!(interpreter, memory_offset);
+    resize_memory!(interpreter, memory_offset, len);
+    if data_offset < return_data_buffer_len {
+        let available_len = return_data_buffer_len - data_offset;
+        let copy_len = available_len.min(len);
+
         interpreter.shared_memory.set(
             memory_offset,
-            &interpreter.return_data_buffer[data_offset..data_end],
+            &interpreter.return_data_buffer[data_offset..data_offset + copy_len],
         );
+
+        if interpreter.is_eof && copy_len < len {
+            interpreter
+                .shared_memory
+                .slice_mut(memory_offset + copy_len, len - copy_len)
+                .fill(0);
+        }
```

Comment:
`else if {error}` is missing. Will restructure code a little bit here.

Reply:
You're right, go ahead 💪